### ISSUE TITLE
Adding guideline for DGE representation

### ISSUE THREAD

**Issue by EstherLerouzic** [ietf-ccamp-wg/draft-ietf-ccamp-optical-impairment-topology-yang#153]
Our draft does not indicate yet how to represent a DGE. There might be various ways to do that, but I would suggest a single one to avoid interpretation. Here is a proposal to put in the draft:

Dynamic gain equalizer (DGE) are equipment pieces that can tune the per channel power to compensate for variable gain or loss history experienced by channels during their propagation. Such an equipment MUST be modelled in the topology using a node without any any tunnel-termination-point. This choice of representation is independent from its hardware realization.

**Comment by sergiobelotti** [comment]
Such an equipment MUST be modelled in the topology using a node with only 2 degrees LTPs .

**Comment by manzoro** [comment]
Optical solution exists on the market that implement DGE functionalities in ILA without actually terminating the OMS link and that are not suitable for the 2 degree LTPs modelling.
This power equalization solution is not based on WSS filtering + VOA and re-muxing but is based on a different technology that we can call Gain shaping or Gain contouring.
It is a concept similar to a GFF(Gain Flattering Filter) but with the capability to be dynamically configured in order to perfectly match (with some approximation) the gain profile required to achieve power equalization (constant PSD over the full spectrum).

Since this equipment do not filter out channels they are not terminating the OMS layer and are more suitable to be modeled as OMS elements with its associated parameters and impairments. 

Then we may need to model the DGE function in different way depending on their HW implementation:
- as a 2-degrees te-node terminating the OMS MCGs (traditional WSS based DGE)
- as a new OMS element, not terminating the OMS MCG (Gain Shaping equalization based DGE)

**Comment by sergiobelotti** [comment]
Call on 10/17/23 
During the call we decided:

- to add agreed text "Dynamic Gain Equalizer (DGE) are pieces of equipment that can tune the per-channel power or the power spectral density to compensate for variable gain or loss history experienced by DWDM signals during their propagation."
- to add attribute in YANG characterize an amplifier element as "DGE" 
- to add PDL at amplifier-element level
- as post meeting notes form Julien and Enrico , we add "delta-power" in the amplifier-element to provide to provide pre-emphasis different from the one provided by the ROADM

This is the new tree created:

```
+--ro amplifier-element* []
    +--ro name?
    |       string
    +--ro is-dynamic-gain-equalyzer?
    |       boolean
    +--ro frequency-range
    |  +--ro lower-frequency    frequency-thz
    |  +--ro upper-frequency    frequency-thz
    +--ro actual-gain
    |       l0-types:gain-in-db-or-null
    +--ro tilt-target
    |       l0-types:decimal-2-digits-or-null
    +--ro out-voa
    |       l0-types:loss-in-db-or-null
    +--ro in-voa
    |       l0-types:loss-in-db-or-null
    +--ro total-output-power
    |       l0-types:power-in-dbm-or-null
    +--ro (power-param)?
    |  +--:(channel-power)
    |  |  +--ro nominal-carrier-power?
    |  |          l0-types:power-in-dbm-or-null
    |  +--:(power-spectral-density)
    |     +--ro nominal-power-spectral-density?
    |             l0-types:decimal-16-digits-or-null
    +--ro raman-direction?
    |       enumeration
    +--ro raman-pump* []
    |  +--ro frequency?
    |  |       l0-types:frequency-thz
    |  +--ro power?
    |          l0-types:decimal-2-digits-or-null
    +--ro pdl?
    |       l0-types:loss-in-db-or-null
    +--ro media-channel-groups
        +--ro media-channel-group* []
          +--ro media-channels* []
              +--ro flexi-n?
              |       l0-types:flexi-n
              +--ro flexi-m?
              |       l0-types:flexi-m
              +--ro delta-power?
                      l0-types:power-in-dbm-or-null

```

**Comment by dieterbeller** [comment]
Revision 14 of the draft will include the following subsection 2.4.1 in section 2.4 Optical Amplifiers:

```
2.4.1.  Dynamic Gain Equalizers

   [Editor's note: Placeholder for DGE subsection paragraphs.]
```

**Comment by manzoro** [comment]
Attaching to the issue the drawing used during last meeting to discuss the option to have the amplifier elements list arrange in order to support both parallel and cascade amplification.

We agreed to have a common mandatory type-variety in the amplifier container, and additional optional type-variety in the amplifier elements.

I think we didn't close on the final position of the is-dynamic-gain-equalizer (leave it in amplifier element as it is now or move it in the main amplifier container)

Will prepare a text proposal with related ascii artwork for the next meeting

[MultiStageAmplifierExample.pptx](https://github.com/ietf-ccamp-wg/draft-ietf-ccamp-optical-impairment-topology-yang/files/13740557/MultiStageAmplifierExample.pptx)

**Comment by manzoro** [comment]
Hi All, please find here the text change proposal to introduce the modelling of multiple amplifier stage:

`
Section 2.4
[...]
The actual-gain attribute reflects the gain of the ILA gain
stage and does not include the attenuation of the in-voa and/or out-
voa.

--- new text ---
In order to support the modelling of multi badwisth (parallel)
and multi stage (cascaded) amplifier like in picture ??, 
the OMS elemnt that describe an optical amplifier contains an unordered 
list of amplifier-element. The position of the elements is bases 
on the following two attributes:

- the lower-frequecy and upper-frequecy attributre indentify the
set of amplifier-element that support a specific spectrum bandwidth
- the stage-order attribute allow to define for each spectrum badwidth
the cascade order of amplifier-element.

<picture like the one shared in MultistageAmplifierExample.pptx>

The deatailed rapresentation of the apmplifier stage is not mandatory in general,
but in the case the OMS-Elemtn has both the funtion of Amplifier and DGE (refer 
section 2.5) the implementation MUST provide the cascade order details of
amplifier element in order to allow a proper modelling of the impairment
--- end new text ---

ILAs are placed at locations where the optical amplification of the
[...]
`

For the model point of view the idea is to add the attribute stage-order in the amplifier-element container:
`
                |  +--ro amplifier
                |     +--ro type-variety    string
                |     +--ro operational
                |        +--ro amplifier-element* []
                |           +--ro name?
                |           |       string
                |           +--ro is-dynamic-gain-equalyzer?
                |           |       boolean
                |           +--ro frequency-range
                |           |  +--ro lower-frequency    frequency-thz
                |           |  +--ro upper-frequency    frequency-thz
                |           +--ro stage-order? int
[...]
`

**Comment by sergiobelotti** [comment]
> Hi All, please find here the text change proposal to introduce the modelling of multiple amplifier stage:
> 
> ```
> Section 2.4
> [...]
> The actual-gain attribute reflects the gain of the ILA gain
> stage and does not include the attenuation of the in-voa and/or out-
> voa.
> 
> --- new text ---
> In order to support the modelling of multi badwisth (parallel)
> and multi stage (cascaded) amplifier like in picture ??, 
> the OMS elemnt that describe an optical amplifier contains an unordered 
> list of amplifier-element. The position of the elements is bases 
> on the following two attributes:
> 
> - the lower-frequecy and upper-frequecy attributre indentify the
> set of amplifier-element that support a specific spectrum bandwidth
> - the stage-order attribute allow to define for each spectrum badwidth
> the cascade order of amplifier-element.
> 
> <picture like the one shared in MultistageAmplifierExample.pptx>
> 
> The deatailed rapresentation of the apmplifier stage is not mandatory in general,
> but in the case the OMS-Elemtn has both the funtion of Amplifier and DGE (refer 
> section 2.5) the implementation MUST provide the cascade order details of
> amplifier element in order to allow a proper modelling of the impairment
> --- end new text ---
> 
> ILAs are placed at locations where the optical amplification of the
> [...]
> ```
> 
> For the model point of view the idea is to add the attribute stage-order in the amplifier-element container:
> 
> ```
>                 |  +--ro amplifier
>                 |     +--ro type-variety    string
>                 |     +--ro operational
>                 |        +--ro amplifier-element* []
>                 |           +--ro name?
>                 |           |       string
>                 |           +--ro is-dynamic-gain-equalyzer?
>                 |           |       boolean
>                 |           +--ro frequency-range
>                 |           |  +--ro lower-frequency    frequency-thz
>                 |           |  +--ro upper-frequency    frequency-thz
>                 |           +--ro stage-order? int
> [...]
> ```
@manzoro : thanks Roberto for the proposal! Apart for some editorial small details I think it is a good text .
About the position of the attribute "is-dynamic-gain-equalizer" I thinks the choice is depending whether we need to characterize the fact one of the "amplifier-element" belonging to the same OMS-element is a DGE e.g. to add specific attributes  related to DGE.  If not , and we need just to make explicit that one OMS-element is supporting DGE functionality, the attributes could be moved above at OMS-element level, in the main amplifier container. 
I've uploaded your pictures adding just the mapping of data nodes model.
[MultiStageAmplifierExample-sb.pptx](https://github.com/ietf-ccamp-wg/draft-ietf-ccamp-optical-impairment-topology-yang/files/13860899/MultiStageAmplifierExample-sb.pptx)

**Comment by egriseri** [comment]
@manzoro, @sergiobelotti
Just for clarification, if we consider the "is-dynamic-gain-equalizer" as an attribute of the amplifier-element, should we make all the other attributes of the amplifier-element as optional (the DGE as amplifier element has no gain or tilt, for example)?

**Comment by EstherLerouzic** [comment]
if we agree that when the attribute is-dynamic-gain-equalizer is present , none of the other mandatory attributes (gain, in_voa, out_voa,...) are present (mutually exclusive) I would prefer a choice on these attributes, to maintain the mandatory. 
(here is a manual made tree of what I have in mind)
`
                |  +--ro amplifier
                |     +--ro type-variety    string
                |     +--ro operational
                |        +--ro amplifier-element* []
                |           +--ro name?
                |           |       string
                |           +--ro frequency-range
                |           |  +--ro lower-frequency    frequency-thz
                |           |  +--ro upper-frequency    frequency-thz
                |           +--ro (power-param)?
                |           |  +--:(channel-power)
                |           |  |  +--ro nominal-carrier-power?
                |           |  |          l0-types:power-in-dbm-or-null
                |           |  +--:(power-spectral-density)
                |           |     +--ro nominal-power-spectral-density?
                |           +--ro stage-order? int
                |           +--ro (amp-or-dge)
                |           |  +--(dge)
                |           |  |  +--is-dynamic-gain-equalyzer     boolean
                |           |  +--(amp)
                |           |  |   +--ro actual-gain     l0-types:gain-in-db-or-null
                |           |  |   +--ro tilt-target    l0-types:decimal-2-digits-or-null
                |           |  |   +--ro out-voa
                |           |  |  +--ro tilt-target
[...]
`

**Comment by italobusi** [comment]
I think with the choice is should become:

```
                |  +--ro amplifier
                |     +--ro type-variety    string
                |     +--ro operational
                |        +--ro amplifier-element* []
                |           +--ro name?
                |           |       string
                |           +--ro frequency-range
                |           |  +--ro lower-frequency    frequency-thz
                |           |  +--ro upper-frequency    frequency-thz
                |           +--ro (power-param)?
                |           |  +--:(channel-power)
                |           |  |  +--ro nominal-carrier-power?
                |           |  |          l0-types:power-in-dbm-or-null
                |           |  +--:(power-spectral-density)
                |           |     +--ro nominal-power-spectral-density?
                |           +--ro stage-order? int
                |           +--ro (amp-or-dge)
                |           |  +--(dge)
                |           |  |  +--ro dynamic-gain-equalyzer?     empty
                |           |  +--(amp)
                |           |  |   +--ro optical-amplifier
                |           |  |   |   +--ro actual-gain     l0-types:gain-in-db-or-null
                |           |  |   |   +--ro tilt-target    l0-types:decimal-2-digits-or-null
                |           |  |   |   +--ro out-voa
                |           |  |   |   +--ro tilt-target
[...] 
```

**Comment by italobusi** [comment]
> in case the amp element is a dge, dynamic-gain-equalizer is mandory in my opinion, otherwise we would allow amps without any settings, which I prefer to avoid. so I am fine with the type, but I prefer to have it mandatory (no ?)

Since the choice is mandatory, either the dynamic-gain-equalyzer leaf or the optical-amplifier container must be present to indicate whether the amplifier element is a DGE or an OA

However, if the media-channel-group applies only to DGE, my suggestion would be to make the dynamic-gain-equalyzer a presence container and to move the media-channel-group within the presence container:

```
                |  +--ro amplifier
                |     +--ro type-variety    string
                |     +--ro operational
                |        +--ro amplifier-element* []
                |           +--ro name?
                |           |       string
                |           +--ro frequency-range
                |           |  +--ro lower-frequency    frequency-thz
                |           |  +--ro upper-frequency    frequency-thz
                |           +--ro (power-param)?
                |           |  +--:(channel-power)
                |           |  |  +--ro nominal-carrier-power?
                |           |  |          l0-types:power-in-dbm-or-null
                |           |  +--:(power-spectral-density)
                |           |     +--ro nominal-power-spectral-density?
                |           +--ro stage-order? int
                |           +--ro (amp-or-dge)
                |           |  +--(dge)
                |           |  |  +--ro dynamic-gain-equalyzer!
                |           |  |     +--ro media-channel-groups
                |           |  |        +--ro media-channel-group* []
                |           |  |           +--ro media-channels* []
                |           |  |              +--ro flexi-n?
                |           |  |              |       l0-types:flexi-n
                |           |  |              +--ro flexi-m?
                |           |  |              |       l0-types:flexi-m
                |           |  |              +--ro delta-power?
                |           |  |                      l0-types:gain-in-db-or-null
                |           |  +--(amp)
                |           |  |   +--ro optical-amplifier
                |           |  |   |   +--ro actual-gain     l0-types:gain-in-db-or-null
                |           |  |   |   +--ro tilt-target    l0-types:decimal-2-digits-or-null
                |           |  |   |   +--ro out-voa
                |           |  |   |   +--ro tilt-target
[...] 
```

The presence container for the dynamic-gain-equalyzer is needed to cover the "corner" case where the amplifier element is a DGE but no MC has yet been configured

The advantage of the external container is that the model can be extended to add new cases with some of the attributes already defined for the DGE or the OA

Maybe, also the name of the choice can be changed from `amp-or-dge` to `amplifier-element-type` to be open to future extensions

My 2 cents

**Comment by italobusi** [comment]
> The deatailed rapresentation of the apmplifier stage is not mandatory in general,
> but in the case the OMS-Elemtn has both the funtion of Amplifier and DGE (refer 
> section 2.5) the implementation MUST provide the cascade order details of
> amplifier element in order to allow a proper modelling of the impairment

Proposed rephrase:

```
The detailed rapresentation of the amplifier stages is not always mandatory, abstraction is allowed
as long as the optical impairments of the multi-stage amplifier are properly modelled.
For example, the detailed representation of the cascade order details can be required in the 
case the OMS-Element has both the function of Amplifier and DGE (see section 2.5).
```

One doubt is about the term 'mandatory' which is a BCP14 keyword: should be capitalize it (adding the proper template text to the I-D) or change it to avoid any comment?

**Comment by egriseri** [comment]
I really like @italobusi proposal.

**Comment by sergiobelotti** [comment]
> the term 'mandatory' which is a BCP14 keyword: should be capitalize it (adding the proper template text to the I-D)

@italobusi: about the term "mandatory", I've checked BCP14 but it seems mandatory is not in the list of keyword.

**Comment by manzoro** [comment]
Updated text proposal with the merge of @italobusi rephrase and with the ascii-art picture
```
Section 2.4
[...]
The actual-gain attribute reflects the gain of the ILA gain
stage and does not include the attenuation of the in-voa and/or out-
voa.

--- new text ---
In order to support the modelling of multi bandwidths (parallel)
and multi stage (cascaded) amplifier like in picture ??, 
the OMS element that describe an optical amplifier contains an unordered 
list of amplifier-element. The position of the elements is based on the 
following two attributes:

- the lower-frequecy and upper-frequecy attributre identify the
set of amplifier-element that support a specific spectrum bandwidth
- the stage-order attribute allow to define for each spectrum bandwidth
the cascade order of amplifier-element.

       +---------------------------------------------+
       |"amplifier" container                        |
       |                                             |
       |                                             |
       |                        BAND C               |
       |                lover/upper-frequency        |
       |                          |                  |
       |              +-----------+----------+       |
       |              |                      |       |
       |                OA1      DGE      OA2        |
       |                |\      +---+     |\         |
       |                | \     |   |     | \        |
    -->+-+------------->|  +----+   +-----+  +-->+---+--->
       | |              | /     |   |     | /    |   |
       | |              |/      +---+     |/     |   |
       | | stage-order = 1        2        3     |   |
       | |                                       |   |
       | |                                       |   |
       | | stage-order = 1        2        3     |   |
       | |              |\      +---+     |\     |   |
       | |              | \     |   |     | \    |   |
       | +------------->|  +----+   +-----+  +-->+   |
       |                | /     |   |     | /        |
       |                |/      +---+     |/         |
       |                OA1      DGE      OA2        |
       |              |                       |      |
       |              +-----------+-----------+      |
       |                          |                  |
       |                lover/upper-frequency        |
       |                       BAND L                |
       |                                             |
       +---------------------------------------------+

Picture ??: Example of dual-band, multi-stage amplifier with DGE functionality 
            OA and DGE blocks modeled as amplifier-element

The detailed representation of the amplifier stages is not always mandatory, abstraction is allowed
as long as the optical impairments of the multi-stage amplifier are properly modeled.
For example, the detailed representation of the cascade order details can be required in the 
case the OMS-Element has both the function of Amplifier and DGE (see section 2.5).
--- end new text ---

ILAs are placed at locations where the optical amplification of the
[...]
```

**Comment by dieterbeller** [comment]
Proposed text incorporated into the I-D with some editorial changes - @manzoro . @italobusi, please have a look!

```
   To support the modeling of multi-band (e.g., C + L band) and multi-
   stage (cascaded) amplifiers as depicted in Figure 6, the OMS element
   that describes an optical amplifier may contain an unordered list of
   amplifier-elements.  The position of the element is based on the
   following two attributes:

   *  lower-frequency and upper-frequency describing the frequency band
      the set of amplifier-elements are operating in.
   *  stage-order describing the sequential order of the cascaded
      amplifier-elements for the frequency band.

   The detailed representation of the amplifier stages is not always
   mandatory.  Abstraction is allowed as long as the optical impairments
   of the multi-stage amplifier are modeled properly.  For example, the
   detailed representation of the cascaded elements is needed in case
   the amplifier supports both amplification of the signal as well as
   the DGE function described in Section 2.5.


                 Dual-band, Multi-stage Amplifier with DGE
             +-----------------------------------------------+
             |                                               |
             |                         C BAND                |
             |                  lower/upper-frequency        |
             |                            |                  |
             |                +-----------+----------+       |
             |                |                      |       |
             |                  OA1      DGE      OA2        |
             |                  |\      +---+     |\         |
             |                  | \     |   |     | \        |
         --->o---+------------->|  +----+   +-----+  +-->+---o--->
             |   |              | /     |   |     | /    |   |
             |   |              |/      +---+     |/     |   |
             |   | stage-order = 1        2        3     |   |
             |   |                                       |   |
             |   |                                       |   |
             |   | stage-order = 1        2        3     |   |
             |   |              |\      +---+     |\     |   |
             |   |              | \     |   |     | \    |   |
             |   +------------->|  +----+   +-----+  +-->+   |
             |                  | /     |   |     | /        |
             |                  |/      +---+     |/         |
             |                  OA1      DGE      OA2        |
             |                |                       |      |
             |                +-----------+-----------+      |
             |                            |                  |
             |                  lower/upper-frequency        |
             |                         L BAND                |
             |                                               |
             +-----------------------------------------------+

      Figure 6: Example of a Dual-band, Multi-stage Amplifier with DGE
                               Functionality

```

**Comment by manzoro** [comment]
@dieterbeller looks great for me! Thanks

**Comment by dieterbeller** [comment]
Please find below a proposal for the additional sentence as agreed during our meeting on Feb 13, 2024. The new sentence is supposed to be added between the last paragraph and the figure (see https://github.com/ietf-ccamp-wg/draft-ietf-ccamp-optical-impairment-topology-yang/issues/153#issuecomment-1935916125 above):

```
   Multi-band amplifiers like the dual-band amplifier depicted in
   Figure 6 have a band-separating filter at the input and a combiner
   combining all the bands at the output.  These functions are not
   modeled explicitly and their optical impairments are subsumed in the
   optical impairments of the amplifier components.
```
This proposal addresses https://github.com/ietf-ccamp-wg/draft-ietf-ccamp-optical-impairment-topology-yang/issues/130#issuecomment-1926587183 from @ggrammel

**Comment by egriseri** [comment]
@dieterbeller I agree with you description. Thanks!

**Comment by sergiobelotti** [comment]
Weekly call 20/02-24:
Multi-band amplifiers like the dual-band amplifier depicted in
   Figure 6 have a band-separating filter at the input and a band combining multiplexer
   combining all the bands at the output.  This filter and multiplexer functions are not
   modeled explicitly and their optical impairments are subsumed in the
   optical impairments of the amplifier components.

**Comment by dieterbeller** [comment]
Paragraph updated (see [c9b7f01](https://github.com/ietf-ccamp-wg/draft-ietf-ccamp-optical-impairment-topology-yang/commit/c9b7f011d183a5189dd3186e7fcfbbd12724fe2e)):

```
   Multi-band amplifiers like the dual-band amplifier depicted in
   Figure 6 have a band-separating filter at the input and a band-
   combining multiplexer combining all the bands at the output.  This
   filter and multiplexer functions are not modeled explicitly and their
   optical impairments are subsumed in the optical impairments of the
   amplifier components.
```

**Comment by sergiobelotti** [comment]
This issue is closed with PR#170
